File: tools/tui/src/lib/auth-session.ts
import { promises as fs } from "node:fs";
import os from "node:os";
import path from "node:path";

export interface AuthSession {
  token: string;
  exp: number | null;
  savedAt: string;
}

const SESSION_DIR = path.join(os.homedir(), ".6flow");
const SESSION_FILE = path.join(SESSION_DIR, "tui-auth.json");

function decodeBase64Url(input: string): string {
  const normalized = input.replace(/-/g, "+").replace(/_/g, "/");
  const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, "=");
  return Buffer.from(padded, "base64").toString("utf8");
}

export function decodeJwtExp(token: string): number | null {
  const parts = token.split(".");
  if (parts.length < 2) return null;

  try {
    const payload = JSON.parse(decodeBase64Url(parts[1])) as { exp?: unknown };
    return typeof payload.exp === "number" ? payload.exp : null;
  } catch {
    return null;
  }
}

export function isSessionValid(session: AuthSession | null): session is AuthSession {
  if (!session || !session.token || typeof session.exp !== "number") {
    return false;
  }

  // 5-second skew buffer.
  return session.exp * 1000 > Date.now() + 5000;
}

export async function loadAuthSession(): Promise<AuthSession | null> {
  try {
    const raw = await fs.readFile(SESSION_FILE, "utf8");
    const parsed = JSON.parse(raw) as Partial<AuthSession>;
    if (typeof parsed.token !== "string") return null;

    return {
      token: parsed.token,
      exp: typeof parsed.exp === "number" ? parsed.exp : decodeJwtExp(parsed.token),
      savedAt:
        typeof parsed.savedAt === "string" ? parsed.savedAt : new Date().toISOString(),
    };
  } catch {
    return null;
  }
}

export async function saveAuthSession(token: string): Promise<AuthSession> {
  const session: AuthSession = {
    token,
    exp: decodeJwtExp(token),
    savedAt: new Date().toISOString(),
  };

  await fs.mkdir(SESSION_DIR, { recursive: true });
  await fs.writeFile(SESSION_FILE, JSON.stringify(session, null, 2), {
    mode: 0o600,
  });
  await fs.chmod(SESSION_FILE, 0o600);

  return session;
}

export async function clearAuthSession(): Promise<void> {
  try {
    await fs.rm(SESSION_FILE, { force: true });
  } catch {
    // No-op.
  }
}

File: tools/tui/internal/tui/auth_session.go
package tui

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type AuthSession struct {
	Token   string `json:"token"`
	Exp     *int64 `json:"exp"`
	SavedAt string `json:"savedAt"`
}

func sessionFilePath() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return ".6flow/tui-auth.json"
	}
	return filepath.Join(home, ".6flow", "tui-auth.json")
}

func decodeJWTExp(token string) *int64 {
	parts := strings.Split(token, ".")
	if len(parts) < 2 {
		return nil
	}

	payloadPart := parts[1]
	if rem := len(payloadPart) % 4; rem != 0 {
		payloadPart += strings.Repeat("=", 4-rem)
	}

	decoded, err := base64.URLEncoding.DecodeString(payloadPart)
	if err != nil {
		return nil
	}

	var payload map[string]any
	if err := json.Unmarshal(decoded, &payload); err != nil {
		return nil
	}

	expFloat, ok := payload["exp"].(float64)
	if !ok {
		return nil
	}

	exp := int64(expFloat)
	return &exp
}

func IsSessionValid(session *AuthSession) bool {
	if session == nil || session.Token == "" || session.Exp == nil {
		return false
	}

	return (*session.Exp)*1000 > time.Now().UnixMilli()+5000
}

func LoadAuthSession() (*AuthSession, error) {
	content, err := os.ReadFile(sessionFilePath())
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	var session AuthSession
	if err := json.Unmarshal(content, &session); err != nil {
		return nil, nil
	}
	if session.Token == "" {
		return nil, nil
	}
	if session.Exp == nil {
		session.Exp = decodeJWTExp(session.Token)
	}
	if session.SavedAt == "" {
		session.SavedAt = time.Now().UTC().Format(time.RFC3339)
	}

	return &session, nil
}

func SaveAuthSession(token string) (*AuthSession, error) {
	exp := decodeJWTExp(token)
	session := &AuthSession{
		Token:   token,
		Exp:     exp,
		SavedAt: time.Now().UTC().Format(time.RFC3339),
	}

	file := sessionFilePath()
	if err := os.MkdirAll(filepath.Dir(file), 0o700); err != nil {
		return nil, err
	}

	content, err := json.MarshalIndent(session, "", "  ")
	if err != nil {
		return nil, err
	}

	if err := os.WriteFile(file, content, 0o600); err != nil {
		return nil, err
	}

	return session, nil
}

func ClearAuthSession() error {
	err := os.Remove(sessionFilePath())
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}
